proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+
+  * A new warning
+    ``-Wmissing-language-edition``, will inform users
+    that they should specify a language edition. This warning will be
+    off by default in GHC, but it is expected that cabal will turn it on.
+    In this way, someone experimenting with a standalone Haskell file doesn't
+    have to fuss about language editions, but anyone building a package
+    (presumably for publication) is encouraged to choose.
+
+  * Starting in 2026, the default language edition when none is specified
+    will be the latest ``Stable`` edition available. (The warning continues,
+    unchanged.)
+
+  * In GHCi, the default language edition will be the latest ``Stable``
+    edition. ``-Wmissing-language-edition`` will be off by default.
+
+- In GHCi, the top-level can have a distinct language edition from loaded
+  files. If a file is being loaded and does not specify its own language
+  edition, it inherits from the language edition in force at the top-level.
+
+- A language edition can control almost all behaviors of GHC. The meaning
+  (or existence) of other flags can depend on language edition. While
+  we will not implement it this way, we can imagine that GHC becomes
+  a set of programs that happen to share a binary; the choice of which
+  program is chosen by the language edition.
+
+  The one restriction on the expressive power of language editions
+  is that build products of different language editions (within the
+  same version of GHC) must be
+  compatible. We expect the Haskell ecosystem to contain packages
+  compiled with a variety of language editions, and they must work
+  together. The word *compatible* above is doing some heavy lifting,
+  in that it contains a notion that the interface (i.e. exported
+  symbols, their types, etc.) can be translated from one language
+  edition to another. This aspect of Haskell's design has always
+  been present, in that different files can be compiled with different
+  langauge extensions. We often don't notice this, because we have
+  been careful with language extensions not to make the translation
+  apparent.
+
+  Although language editions have wide authority, we must be tasteful
+  in how they work. It would be problematic to have ``Stable2024``
+  and ``Stable2027`` disagree on the meaning of widely used features.
+  Yet I think it's best to be maximally expressive here, relying on
+  our future selves not to abuse our power.
+
+  (In the conversation, it was suggested that a ``Python`` language
+  edition would be beyond reason... but actually I think such a thing
+  would be lovely, if the necessary work was done to translate the
+  interfaces between Python and Haskell.)
+
+- For backward compatibility, a language edition can be specified
+  at an arbitrary place in a command-line invocation of GHC, or in a
+  ``LANGUAGE`` pragma in a file. Even though it might come later in
+  a command line, the edition can affect the meaning of command-line
+  arguments that precede it.
```

Comment:
On reflection, I think it probably is fine to allow later specification of editions, we'll just need to "bubble them up" before starting flag processing.

Reply:
Yes, that's the intent. Slurp in all the flags first, then identify the language edition, then process the flags. I think this will be fine in practice.